nixos/matrix-synapse: mergeable log configuration

Ma27 opened the pull request:
## Description of changes
Right now there's no trivial way to override parts of synapse's log config such as the log-level because the only thing that's changeable is the path to the log-file used by synapse and its workers.

Now, there's a new option called `services.matrix-synapse.log` which contains the default log config as Nix attribute-set (except `handlers.journal.SYSLOG_IDENTIFIER`). It has default priority, so new things can be added like

    services.matrix-synapse.log = {
      my.extra.field = 23;
    }

without discarding the rest. If desired, this can still be done via `lib.mkForce`.

If the log configuration for a single worker or synapse, but not all workers should be changed, `services.matrix-synapse.settings.log_config` or `services.matrix-synapse.workers._name_.worker_log_config` can be used.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Ma27 committed:
nixos/matrix-synapse: mergeable log configuration

Right now there's no trivial way to override parts of synapse's log
config such as the log-level because the only thing that's changeable is
the path to the log-file used by synapse and its workers.

Now, there's a new option called `services.matrix-synapse.log`
which contains the default log config as Nix attribute-set (except
`handlers.journal.SYSLOG_IDENTIFIER`). It has default priority, so new
things can be added like

    services.matrix-synapse.log = {
      my.extra.field = 23;
    }

without discarding the rest. If desired, this can still be done via
`lib.mkForce`.

If the log configuration for a single worker or synapse, but not all
workers should be changed, `services.matrix-synapse.settings.log_config`
or `services.matrix-synapse.workers._name_.worker_log_config`
can be used.

SuperSandro2000 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -394,6 +387,47 @@ in {
+        defaultText = literalExpression defaultCommonLogConfigText;
+        description = mdDoc ''
+          Default configuration for the loggers used by `matrix-synapse` and its workers.
+          The defaults are added with the default priority which means that
+          these will be merged with additional declarations. For instance
```
I am confused a bit because default priority would mean that I wouldn't need to use mkForce.

Ma27 replied:
No. Using option priority (`mkOption { default = defaultLogConfig; }`) means that declaring _anything_ in `services.matrix-synapse.log` _discards_ everything from `default`.

Setting everything with default priority means that you can add additional properties, but you'll need a higher priority (`mkForce`) to override existing declarations.

SuperSandro2000 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -1008,6 +1042,8 @@ in {
     # default them, so they are additive
     services.matrix-synapse.extras = defaultExtras;
 
+    services.matrix-synapse.log = defaultCommonLogConfig;
```
If we would use something like, then mkForce wouldn't be required.

```suggestion
    services.matrix-synapse.log = mapAttrsRecursive (_: lib.mkDefault) defaultCommonLogConfig;
```

Ma27 replied:
Yeah, I considered doing something like that, but I'm not sure if that's too much magic (it would be less of an issue if we didn't nested attribute sets here IMHO).

What do the others think?

999eagle replied:
I think making this recursively default makes it a lot harder for people who want to override the log config in its entirety as every property would need to be unset/overridden. I'm not sure though whether there's actually a useful use case for this, so unless we find any I'm fine with the proposed change to `mkDefault`.

Ma27 replied:
> I think making this recursively default makes it a lot harder for people who want to override the log config in its entirety as every property would need to be unset/overridden

This should be doable with `services.matrix-synapse.log = mkForce ...;`.

The sole difference of what @SuperSandro2000 suggested should be whether you need to override e.g. `services.matrix-synapse.log.root.level` with `mkForce` (currently the case with my patch) or not (doable when recursively doing `mkDefault`).

However, revisiting this again: I was concerned about corner cases (where it's more correct to `mkDefault` an attribute-set rather than its attribtues inside), but we know the default, i.e. the configuration where `mapAttrsRecursive` gets applied upon and that one seems fine, so I'm inclined to just do that.

SuperSandro2000 replied:
I think I am fine with both options :)

Ma27 committed:
nixos/synapse: `mkDefault` each value of the log config

That way it's not even needed to specify an `mkForce` when changing
existing attributes, e.g. root's log level.
